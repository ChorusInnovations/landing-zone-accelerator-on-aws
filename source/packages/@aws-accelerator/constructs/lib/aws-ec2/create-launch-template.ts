/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as fs from 'fs';

export interface ILaunchTemplateResource extends cdk.IResource {
  /**
   * The version number.
   */
  readonly version: string;
  /**
   * The ID of the AWS::EC2::LaunchTemplate . You must specify either a LaunchTemplateName or a LaunchTemplateId .
   */
  readonly launchTemplateId: string;
  /**
   * The name of the AWS::EC2::LaunchTemplate . You must specify either a LaunchTemplateName or a LaunchTemplateId .
   */
  readonly launchTemplateName: string;
}
export type PrivateIpAddressConfig = {
  primary: boolean | undefined;
  privateIpAddress: string | undefined;
};
export type NetworkInterfaceItemConfig = {
  associateCarrierIpAddress: boolean | undefined;
  associatePublicIpAddress: boolean | undefined;
  deleteOnTermination: boolean | undefined;
  description: string | undefined;
  deviceIndex: number | undefined;
  groups: string[] | undefined;
  interfaceType: string | undefined;
  networkCardIndex: number | undefined;
  networkInterfaceId: string | undefined;
  privateIpAddress: string | undefined;
  secondaryPrivateIpAddressCount: number | undefined;
  subnetId: string | undefined;
  privateIpAddresses: PrivateIpAddressConfig[] | undefined;
};
export type EbsProperty = {
  deleteOnTermination?: boolean;
  encrypted?: boolean;
  iops?: number;
  kmsKeyId?: string;
  snapshotId?: string;
  throughput?: number;
  volumeSize?: number;
  volumeType?: string;
};

export type BlockDeviceMappingItem = {
  deviceName: string;
  ebs?: EbsProperty;
};

export interface LaunchTemplateProps {
  /*
   * Path to user data.
   */
  readonly userData?: string;
  /*
   * VpcName
   */
  readonly vpc: string;
  /*
   * Name of Launch Template
   */
  readonly name: string;
  /*
   * Name of Application
   */
  readonly appName: string;
  readonly blockDeviceMappings?: BlockDeviceMappingItem[];
  readonly securityGroups?: string[];
  readonly networkInterfaces?: NetworkInterfaceItemConfig[];
  readonly instanceType: string;
  readonly keyPair?: string;
  readonly iamInstanceProfile?: string;
  readonly imageId: string;
  readonly enforceImdsv2?: boolean;
}

export class LaunchTemplate extends cdk.Resource implements ILaunchTemplateResource {
  public readonly version: string;
  public readonly launchTemplateId: string;
  public readonly launchTemplateName: string;

  constructor(scope: Construct, id: string, props: LaunchTemplateProps) {
    super(scope, id);

    let metadataOptions: cdk.aws_ec2.CfnLaunchTemplate.MetadataOptionsProperty;

    if (props.enforceImdsv2 === false) {
      metadataOptions = { httpTokens: 'optional' };
    } else {
      metadataOptions = { httpTokens: 'required' };
    }

    const launchTemplate = new cdk.aws_ec2.CfnLaunchTemplate(this, 'LaunchTemplate', {
      launchTemplateData: {
        blockDeviceMappings: props.blockDeviceMappings ?? undefined,
        securityGroupIds: props.securityGroups ?? undefined,
        networkInterfaces: props.networkInterfaces ?? undefined,
        instanceType: props.instanceType,
        keyName: props.keyPair ?? undefined,
        imageId: props.imageId,
        iamInstanceProfile: { name: props.iamInstanceProfile ?? undefined },
        userData: cdk.Fn.base64(fs.readFileSync(props.userData!, 'utf8')) ?? undefined,
        metadataOptions,
      },
      launchTemplateName: props.name,
    });

    this.launchTemplateId = launchTemplate.ref;
    this.launchTemplateName = launchTemplate.launchTemplateName!;
    this.version = launchTemplate.attrLatestVersionNumber;
  }
}
