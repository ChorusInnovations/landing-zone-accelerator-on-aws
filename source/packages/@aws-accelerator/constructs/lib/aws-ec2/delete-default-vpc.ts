/**
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { v4 as uuidv4 } from 'uuid';
import { Construct } from 'constructs';

const path = require('path');

/**
 * Class Delete the Default VPC
 */
export class DeleteDefaultVpc extends Construct {
  readonly id: string;
  constructor(scope: Construct, id: string) {
    super(scope, id);

    const DELETE_DEFAULT_VPC_TYPE = 'Custom::DeleteDefaultVpc';

    //
    // Function definition for the custom resource
    //
    const cr = cdk.CustomResourceProvider.getOrCreateProvider(this, DELETE_DEFAULT_VPC_TYPE, {
      codeDirectory: path.join(__dirname, 'delete-default-vpc/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_14_X,
      policyStatements: [
        {
          Effect: 'Allow',
          Action: [
            'ec2:DeleteInternetGateway',
            'ec2:DetachInternetGateway',
            'ec2:DeleteNetworkAcl',
            'ec2:DeleteRoute',
            'ec2:DeleteSecurityGroup',
            'ec2:DeleteSubnet',
            'ec2:DeleteVpc',
            'ec2:DescribeInternetGateways',
            'ec2:DescribeNetworkAcls',
            'ec2:DescribeRouteTables',
            'ec2:DescribeSecurityGroups',
            'ec2:DescribeSubnets',
            'ec2:DescribeVpcs',
          ],
          Resource: '*',
        },
      ],
    });

    //
    // Custom Resource definition. We want this resource to be evaluated on
    // every CloudFormation update, so we generate a new uuid to force
    // re-evaluation.
    //
    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: DELETE_DEFAULT_VPC_TYPE,
      serviceToken: cr.serviceToken,
      properties: {
        uuid: uuidv4(),
      },
    });

    this.id = resource.ref;
  }
}
