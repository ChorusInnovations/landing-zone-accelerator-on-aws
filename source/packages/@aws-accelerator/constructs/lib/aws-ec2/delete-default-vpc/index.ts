/**
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { throttlingBackOff } from '@aws-accelerator/utils';
import * as console from 'console';
import {
  DeleteInternetGatewayCommand,
  DetachInternetGatewayCommand,
  DeleteNetworkAclCommand,
  DeleteRouteCommand,
  DeleteSecurityGroupCommand,
  DeleteSubnetCommand,
  DeleteVpcCommand,
  EC2Client,
  paginateDescribeInternetGateways,
  paginateDescribeNetworkAcls,
  paginateDescribeRouteTables,
  paginateDescribeSecurityGroups,
  paginateDescribeSubnets,
  paginateDescribeVpcs,
} from '@aws-sdk/client-ec2';

/**
 * delete-default-vpc - lambda handler
 *
 * @param event
 * @returns
 */

export async function handler(event: AWSLambda.CloudFormationCustomResourceEvent): Promise<
  | {
    PhysicalResourceId: string | undefined;
    Status: string;
  }
  | undefined
> {
  // Retrieve operating region that stack is ran
  const region = event.ResourceProperties['region'];
  const ec2Client = new EC2Client({ region: region });
  const defaultVpcIds: string[] = [];

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      console.log(`Starting - Deletion of default VPC and associated resources in ${region}`);

      // Retrieve default VPC(s)
      for await (const page of paginateDescribeVpcs(
        { client: ec2Client },
        { Filters: [{ Name: 'is-default', Values: [`true`] }] },
      )) {
        for (const vpc of page.Vpcs ?? []) {
          if (vpc.VpcId) {
            defaultVpcIds.push(vpc.VpcId);
          }
        }
      }
      console.log(`List of VPCs: `, defaultVpcIds);
      if (defaultVpcIds.length == 0) {
        console.warn('No default VPCs detected');
        return {
          PhysicalResourceId: `delete-default-vpc`,
          Status: 'SUCCESS',
        };
      } else {
        console.warn('Default VPC Detected');
      }

      // Retrieve and detach, delete IGWs
      for (const vpcId of defaultVpcIds) {
        for await (const page of paginateDescribeInternetGateways(
          { client: ec2Client },
          { Filters: [{ Name: 'attachment.vpc-id', Values: [vpcId] }] },
        )) {
          for (const igw of page.InternetGateways ?? []) {
            for (const attachment of igw.Attachments ?? []) {
              if (attachment.State == 'available') {
                console.log(`Detaching ${igw.InternetGatewayId}`);
                await throttlingBackOff(() =>
                  ec2Client.send(
                    new DetachInternetGatewayCommand({ InternetGatewayId: igw.InternetGatewayId, VpcId: vpcId }),
                  ),
                );
              }
              console.warn(`${igw.InternetGatewayId} is not attached. Proceeding to delete.`);
              await throttlingBackOff(() =>
                ec2Client.send(
                  new DeleteInternetGatewayCommand({
                    InternetGatewayId: igw.InternetGatewayId,
                  }),
                ),
              );
            }
          }
        }

        // Retrieve Default VPC Subnets
        console.log(`Gathering Subnets for VPC ${vpcId}`);
        for await (const page of paginateDescribeSubnets(
          { client: ec2Client },
          { Filters: [{ Name: 'vpc-id', Values: [vpcId] }] },
        )) {
          for (const subnet of page.Subnets ?? []) {
            console.log(`Delete Subnet ${subnet.SubnetId}`);
            await throttlingBackOff(() =>
              ec2Client.send(
                new DeleteSubnetCommand({
                  SubnetId: subnet.SubnetId,
                }),
              ),
            );
          }
        }

        // Delete Routes
        console.log(`Gathering list of Route Tables for VPC ${vpcId}`);
        for await (const page of paginateDescribeRouteTables(
          { client: ec2Client },
          { Filters: [{ Name: 'vpc-id', Values: [vpcId] }] },
        )) {
          for (const routeTableObject of page.RouteTables ?? []) {
            for (const routes of routeTableObject.Routes ?? []) {
              if (routes.GatewayId !== 'local') {
                console.log(`Removing route ${routes.DestinationCidrBlock} from ${routeTableObject.RouteTableId}`);
                await throttlingBackOff(() =>
                  ec2Client.send(
                    new DeleteRouteCommand({
                      RouteTableId: routeTableObject.RouteTableId,
                      DestinationCidrBlock: routes.DestinationCidrBlock,
                    }),
                  ),
                );
              }
            }
          }
        }

        // List and Delete NACLs
        console.log(`Gathering list of NACLs for VPC ${vpcId}`);
        for await (const page of paginateDescribeNetworkAcls(
          { client: ec2Client },
          { Filters: [{ Name: 'vpc-id', Values: [vpcId] }] },
        )) {
          for (const networkAclObject of page.NetworkAcls ?? []) {
            if (networkAclObject.IsDefault !== true) {
              console.log(`Deleting Network ACL ID ${networkAclObject.NetworkAclId}`);
              await throttlingBackOff(() =>
                ec2Client.send(
                  new DeleteNetworkAclCommand({
                    NetworkAclId: networkAclObject.NetworkAclId,
                  }),
                ),
              );
            } else {
              console.warn(`${networkAclObject.NetworkAclId} is the default NACL. Ignoring`);
            }
          }
        }

        // List and Delete Security Groups
        console.log(`Gathering list of Security Groups for VPC ${vpcId}`);
        for await (const page of paginateDescribeSecurityGroups(
          { client: ec2Client },
          { Filters: [{ Name: 'vpc-id', Values: [vpcId] }] },
        )) {
          for (const securityGroupObject of page.SecurityGroups ?? []) {
            if (securityGroupObject.GroupName == 'default') {
              console.warn(`${securityGroupObject.GroupId} is the default SG. Ignoring`);
            } else {
              console.log(`Deleting Security Group Id ${securityGroupObject.GroupId}`);
              await throttlingBackOff(() =>
                ec2Client.send(
                  new DeleteSecurityGroupCommand({
                    GroupId: securityGroupObject.GroupId,
                  }),
                ),
              );
            }
          }
        }
        // Once all resources are deleted, delete the VPC.
        console.log(`Deleting VPC ${vpcId}`);
        await throttlingBackOff(() => ec2Client.send(new DeleteVpcCommand({ VpcId: vpcId })));
      }

      return {
        PhysicalResourceId: `delete-default-vpc`,
        Status: 'SUCCESS',
      };

    case 'Delete':
      // Do Nothing
      return {
        PhysicalResourceId: event.PhysicalResourceId,
        Status: 'SUCCESS',
      };
  }
}
