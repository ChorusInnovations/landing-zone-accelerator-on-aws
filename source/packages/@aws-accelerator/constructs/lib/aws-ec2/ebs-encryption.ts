/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

const path = require('path');

/**
 * Initialized EbsVolumeEncryptionProps properties
 */
export interface EbsVolumeEncryptionProps {
  readonly kmsKey: cdk.aws_kms.IKey;
}

/**
/**
* Class to Enable Default EBS Volume Encryption
*/
export class EbsDefaultEncryption extends Construct {
  public readonly id: string;

  constructor(scope: Construct, id: string, props: EbsVolumeEncryptionProps) {
    super(scope, id);

    const EBS_ENCRYPTION_TYPE = 'Custom::EbsDefaultVolumeEncryption';

    const iamPolicy = [
      {
        Sid: 'EC2',
        Effect: 'Allow',
        Action: [
          'ec2:DisableEbsEncryptionByDefault',
          'ec2:EnableEbsEncryptionByDefault',
          'ec2:ModifyEbsDefaultKmsKeyId',
          'ec2:ResetEbsDefaultKmsKeyId',
        ],
        Resource: '*',
      },
      {
        Sid: 'KMS',
        Effect: 'Allow',
        Action: ['kms:DescribeKey'],
        Resource: props.kmsKey.keyArn,
      },
    ];

    const enableEncryptionFunction = cdk.CustomResourceProvider.getOrCreateProvider(this, EBS_ENCRYPTION_TYPE, {
      codeDirectory: path.join(__dirname, 'ebs-default-encryption/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_14_X,
      policyStatements: iamPolicy,
    });

    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: EBS_ENCRYPTION_TYPE,
      serviceToken: enableEncryptionFunction.serviceToken,
      properties: {
        kmsKeyId: props.kmsKey?.keyId,
      },
    });

    this.id = resource.ref;
  }
}
