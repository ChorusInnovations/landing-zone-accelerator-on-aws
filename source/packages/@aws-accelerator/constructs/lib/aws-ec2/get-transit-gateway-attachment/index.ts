/**
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { throttlingBackOff } from '@aws-accelerator/utils';
import { EC2Client, paginateDescribeTransitGatewayAttachments } from '@aws-sdk/client-ec2';
import { AssumeRoleCommand, STSClient } from '@aws-sdk/client-sts';

/**
 * get-transit-gateway-attachment - lambda handler
 *
 * @param event
 * @returns
 */
export async function handler(event: AWSLambda.CloudFormationCustomResourceEvent): Promise<
  | {
      PhysicalResourceId: string | undefined;
      Status: string;
    }
  | undefined
> {
  const name = event.ResourceProperties['name'];
  const transitGatewayId = event.ResourceProperties['transitGatewayId'];
  const roleArn = event.ResourceProperties['roleArn'];

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      const stsClient = new STSClient({});

      const assumeRoleResponse = await throttlingBackOff(() =>
        stsClient.send(
          new AssumeRoleCommand({
            RoleArn: roleArn,
            RoleSessionName: 'GetTransitGatewayAttachmentSession',
          }),
        ),
      );

      const ec2Client = new EC2Client({
        credentials: {
          accessKeyId: assumeRoleResponse.Credentials?.AccessKeyId ?? '',
          secretAccessKey: assumeRoleResponse.Credentials?.SecretAccessKey ?? '',
          sessionToken: assumeRoleResponse.Credentials?.SessionToken,
        },
      });

      for await (const page of paginateDescribeTransitGatewayAttachments({ client: ec2Client }, {})) {
        for (const attachment of page.TransitGatewayAttachments ?? []) {
          if (attachment.TransitGatewayId === transitGatewayId) {
            const nameTag = attachment.Tags?.find(t => t.Key === 'Name');
            if (nameTag && nameTag.Value === name) {
              console.log(attachment);
              return {
                PhysicalResourceId: attachment.TransitGatewayAttachmentId,
                Status: 'SUCCESS',
              };
            }
          }
        }
      }

      throw new Error(`Attachment ${name} for ${transitGatewayId} not found`);

    case 'Delete':
      // Do Nothing
      return {
        PhysicalResourceId: event.PhysicalResourceId,
        Status: 'SUCCESS',
      };
  }
}
