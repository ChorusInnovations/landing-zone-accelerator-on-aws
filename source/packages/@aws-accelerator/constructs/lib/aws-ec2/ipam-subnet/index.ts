/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import * as AWS from 'aws-sdk';
import { IPv4Prefix, Pool } from 'ip-num';

import { throttlingBackOff } from '@aws-accelerator/utils';

import { Vpc, vpcInit } from './vpc';

const ec2 = new AWS.EC2();
export async function handler(event: AWSLambda.CloudFormationCustomResourceEvent): Promise<
  | {
      PhysicalResourceId: string;
      Status: string;
      Data: {
        ipv4CidrBlock: string | undefined;
      };
    }
  | {
      PhysicalResourceId: string;
      Status: string;
    }
  | undefined
> {
  // Create interfaces
  interface IpamAllocation {
    readonly ipamPoolName: string;
    readonly netmaskLength: number;
  }
  interface Tag {
    readonly Key: string;
    readonly Value: string;
  }

  // Set properties
  const az: string = event.ResourceProperties['availabilityZone'];
  const basePool: string[] = event.ResourceProperties['basePool'];
  const ipamAllocation: IpamAllocation = event.ResourceProperties['ipamAllocation'];
  let mapPublicIpOnLaunch: boolean | undefined = event.ResourceProperties['mapPublicIpOnLaunch'] ?? false;
  const subnetName: string = event.ResourceProperties['name'];
  const tags: Tag[] = event.ResourceProperties['tags'];
  const vpcId: string = event.ResourceProperties['vpcId'];

  // Set vpc variable
  let vpc: Vpc;

  // Set netmask length as BigInt
  const netmaskLength = BigInt(ipamAllocation.netmaskLength);

  // Handle case where boolean is passed as string
  if (mapPublicIpOnLaunch) {
    mapPublicIpOnLaunch = returnBoolean(mapPublicIpOnLaunch.toString());
  }

  // Handle the custom resource workflow
  switch (event.RequestType) {
    case 'Create':
      // Initialize the VPC
      vpc = await vpcInit({ basePool, vpcId });
      // Get the subnet CIDR
      const subnetCidr = nextCidr(netmaskLength, vpc);

      // Set name tag
      tags.push({ Key: 'Name', Value: subnetName });

      // Create the subnet
      console.log(`Create subnet ${subnetName} with CIDR ${subnetCidr} in VPC ${vpc.name} (${vpc.vpcId})`);
      const response = await throttlingBackOff(() =>
        ec2
          .createSubnet({
            TagSpecifications: [{ ResourceType: 'subnet', Tags: tags }],
            AvailabilityZone: az,
            CidrBlock: subnetCidr,
            VpcId: vpc.vpcId,
          })
          .promise(),
      );
      const subnetId = response.Subnet?.SubnetId;

      if (!subnetId) {
        throw new Error(`Unable to retrieve subnet ID for newly-created subnet`);
      }

      if (mapPublicIpOnLaunch) {
        await throttlingBackOff(() =>
          ec2.modifySubnetAttribute({ MapPublicIpOnLaunch: { Value: true }, SubnetId: subnetId }).promise(),
        );
      }

      return {
        PhysicalResourceId: subnetId,
        Status: 'SUCCESS',
        Data: {
          ipv4CidrBlock: subnetCidr,
        },
      };

    case 'Update':
      // Initialize the VPC
      vpc = await vpcInit({ basePool, vpcId });
      // Get existing subnet from subnet array
      const subnet = vpc.subnets.find(item => item.name === subnetName);

      if (!subnet) {
        throw new Error(`Unable to locate existing subnet ${subnetName}`);
      }

      // Throw error if CIDR prefix value has changed
      if (subnet.allocatedCidr.cidrPrefix.getValue() !== netmaskLength) {
        throw new Error(
          `Cannot allocate new CIDR for existing subnet ${subnetName} (${subnet.subnetId}). Please delete and recreate the subnet instead.`,
        );
      }

      // Update subnet tags and attributes as needed
      if (tags.length > 0) {
        console.log(`Update tags for subnet ${subnetName} (${subnet.subnetId})`);
        await throttlingBackOff(() => ec2.createTags({ Resources: [subnet.subnetId], Tags: tags }).promise());
      }

      if (subnet.mapPublicIpOnLaunch !== mapPublicIpOnLaunch) {
        const value = mapPublicIpOnLaunch ?? false;

        console.log(`Update MapPublicIpOnLaunch property of subnet ${subnetName} (${subnet.subnetId}) to ${value}`);
        await throttlingBackOff(() =>
          ec2.modifySubnetAttribute({ MapPublicIpOnLaunch: { Value: value }, SubnetId: subnet.subnetId }).promise(),
        );
      }

      return {
        PhysicalResourceId: subnet.subnetId,
        Status: 'SUCCESS',
        Data: {
          ipv4CidrBlock: subnet.allocatedCidr.toCidrString(),
        },
      };

    case 'Delete':
      console.log(`Delete subnet ${subnetName} (${event.PhysicalResourceId})`);
      await throttlingBackOff(() => ec2.deleteSubnet({ SubnetId: event.PhysicalResourceId }).promise());

      return {
        PhysicalResourceId: event.PhysicalResourceId,
        Status: 'SUCCESS',
      };
  }
}

function nextCidr(netmaskLength: bigint, vpc: Vpc): string {
  // Instantiate the pool
  const pool = Pool.fromCidrRanges(vpc.allocatedCidrs);

  // Remove existing CIDRs from pool
  for (const subnet of vpc.subnets) {
    const removeCidr = pool.removeOverlapping(subnet.allocatedCidr.toRangeSet());

    if (!removeCidr) {
      throw new Error(
        `Unable to remove existing subnet CIDR ${subnet.allocatedCidr.toCidrString()} from available CIDR pool`,
      );
    }
  }

  // Return next CIDR
  const prefix = IPv4Prefix.fromNumber(netmaskLength);
  return pool.getCidrRange(prefix).toCidrString();
}

function returnBoolean(input: string): boolean | undefined {
  try {
    return JSON.parse(input.toLowerCase());
  } catch (e) {
    return undefined;
  }
}
