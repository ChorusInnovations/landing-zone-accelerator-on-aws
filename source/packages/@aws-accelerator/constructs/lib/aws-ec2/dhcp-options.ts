/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { CfnTag, IResource, Resource } from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';

export interface IDhcpOptions extends IResource {
  /**
   * The name of the DHCP options set.
   */
  readonly name: string;

  /**
   * The ID of the DHCP options set.
   */
  readonly dhcpOptionsId: string;
}

export interface DhcpOptionsProps {
  /**
   * The name of the DHCP options set.
   */
  readonly name: string;

  /**
   * This value is used to complete unqualified DNS hostnames.
   */
  readonly domainName?: string;

  /**
   * The IPv4 addresses of up to four domain name servers.
   *
   * @default -- AmazonProvidedDNS
   */
  readonly domainNameServers?: string[];

  /**
   * The IPv4 addresses of up to four NetBIOS name servers.
   */
  readonly netbiosNameServers?: string[];

  /**
   * The NetBIOS node type (1, 2, 4, or 8).
   */
  readonly netbiosNodeType?: number;

  /**
   * The IPv4 addresses of up to four Network Time Protocol (NTP) servers.
   */
  readonly ntpServers?: string[];

  /**
   * Any tags assigned to the DHCP options set.
   */
  readonly tags?: CfnTag[];
}

export class DhcpOptions extends Resource implements IDhcpOptions {
  public readonly name: string;
  public readonly dhcpOptionsId: string;

  constructor(scope: Construct, id: string, props: DhcpOptionsProps) {
    super(scope, id);

    this.name = props.name;

    // Add name tag to tags
    props.tags?.push({ key: 'Name', value: this.name });

    const resource = new ec2.CfnDHCPOptions(this, 'Resource', {
      domainName: props.domainName,
      domainNameServers: props.domainNameServers ?? ['AmazonProvidedDNS'],
      netbiosNameServers: props.netbiosNameServers,
      netbiosNodeType: props.netbiosNodeType,
      ntpServers: props.ntpServers,
      tags: props.tags,
    });

    this.dhcpOptionsId = resource.attrDhcpOptionsId;
  }
}
