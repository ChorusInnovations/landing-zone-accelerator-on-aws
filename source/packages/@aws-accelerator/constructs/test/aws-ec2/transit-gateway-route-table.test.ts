/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { SynthUtils } from '@aws-cdk/assert';
import { TransitGatewayRouteTable } from '../../lib/aws-ec2/transit-gateway-route-table';

const testNamePrefix = 'Construct(TransitGatewayRouteTable): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

new TransitGatewayRouteTable(stack, 'TransitGatewayRouteTable', {
  name: 'core',
  transitGatewayId: 'tgw0001',
  tags: [{ key: 'Test-Key', value: 'Test-Value' }],
});
/**
 * TransitGatewayRouteTable construct test
 */
describe('TransitGatewayRouteTable', () => {
  /**
   * Snapshot test
   */
  test(`${testNamePrefix} Snapshot Test`, () => {
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
  /**
   * Number of TransitGatewayRouteTable resource test
   */
  test(`${testNamePrefix} TransitGatewayRouteTable resource count test`, () => {
    cdk.assertions.Template.fromStack(stack).resourceCountIs('AWS::EC2::TransitGatewayRouteTable', 1);
  });

  /**
   * TransitGatewayRouteTable resource configuration test
   */
  test(`${testNamePrefix} TransitGatewayRouteTable resource configuration test`, () => {
    cdk.assertions.Template.fromStack(stack).templateMatches({
      Resources: {
        TransitGatewayRouteTableCoreTransitGatewayRouteTableD6BC94E0: {
          Type: 'AWS::EC2::TransitGatewayRouteTable',
          Properties: {
            Tags: [
              {
                Key: 'Name',
                Value: 'core',
              },
              {
                Key: 'Test-Key',
                Value: 'Test-Value',
              },
            ],
            TransitGatewayId: 'tgw0001',
          },
        },
      },
    });
  });
});
