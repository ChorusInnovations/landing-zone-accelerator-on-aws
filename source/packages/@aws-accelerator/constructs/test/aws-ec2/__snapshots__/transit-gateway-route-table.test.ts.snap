// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransitGatewayRouteTable Construct(TransitGatewayRouteTable):  Snapshot Test 1`] = `
Object {
  "Resources": Object {
    "TransitGatewayRouteTableCoreTransitGatewayRouteTableD6BC94E0": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "core",
          },
        ],
        "TransitGatewayId": "tgw0001",
      },
      "Type": "AWS::EC2::TransitGatewayRouteTable",
    },
  },
}
`;
